plugins/woocommerce/src/Internal/DependencyManagement/ServiceProviders/AbstractInterfaceServiceProvider.php

```diff
@@ -0,0 +1,80 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\Internal\DependencyManagement\ServiceProviders;
+
+use Automattic\WooCommerce\Internal\DependencyManagement\AbstractServiceProvider;
+use Automattic\WooCommerce\Vendor\League\Container\Definition\DefinitionInterface;
+
+/**
+ * Extends AbstractServiceProvider to register services and automatically tag them based on their implemented interfaces.
+ * By using the `add_with_implements_tags` and `share_with_implements_tags` methods, it becomes possible to retrieve
+ * all the services that implement a given interface with a single `get` call.
+ *
+ * @since x.x.x
+ */
+abstract class AbstractInterfaceServiceProvider extends AbstractServiceProvider {
+
+	/**
+	 * Determine whether this service provides the given alias.
+	 *
+	 * @param string $alias The alias to check.
+	 *
+	 * @return bool
+	 */
+	public function provides( string $alias ): bool {
+		$provides = parent::provides( $alias );
+		if ( $provides ) {
+			return true;
+		}
+
+		static $implements = array();
+		if ( empty( $implements ) ) {
+			foreach ( $this->provides as $class ) {
+				$implements = array_merge( $implements, class_implements( $class ) );
```

Comment:
`class_implements` may return `false`, which makes all the following code fail, as `$implements` becomes `null`.

I'm not sure why it iterates over classes that are not implemented. For me, it was failing with 
`$class=="Automattic\WooCommerce\Internal\Integrations\WPConsentAPI"` and `$service == "Automattic\WooCommerce\Internal\DependencyManagement\ServiceProviders\ObjectCacheServiceProvider"`

When I added another clone of existing provider
```php
	protected $provides = array(
		SourceAttributionController::class,
		SourceAttributionClonedController::class,
```


```suggestion
				$implements_more = class_implements( $class );
				if( $implements_more ) {
					$implements = array_merge( $implements, $implements_more );
				}
```